src/utils/TodosContext.tsx

```diff
@@ -0,0 +1,199 @@
+  updateTitle: () => {},
+  toggleOne: () => {},
+  toggleAll: () => {},
+  filter: Filter.All,
+  setFilter: () => {},
+  query: '',
+  setQuery: () => {},
+  error: Error.NO_ERROR,
+  setError: () => {},
+  isLoading: false,
+  setIsLoading: () => {},
+  tempTodo: null,
+  setTempTodo: () => {},
+  loadingTodosIds: [],
+  setLoadingTodosIds: () => {},
+});
+
+type Props = {
+  children: ReactNode;
+};
+
+export const TodosProvider: FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filter, setFilter] = useState(Filter.All);
+  const [query, setQuery] = useState('');
+  const [error, setError] = useState(Error.NO_ERROR);
+  const [isLoading, setIsLoading] = useState(false);
+  const [loadingTodosIds, setLoadingTodosIds] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const addTodo = async (newTodo: Omit<Todo, 'id'>) => {
+    setIsLoading(true);
+    setLoadingTodosIds(prev => [...prev, 0]);
+
+    try {
+      const res = await postTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, res]);
+      setQuery('');
+    } catch {
+      setError(Error.UNABLE_TO_ADD);
+    } finally {
+      setTempTodo(null);
+      setIsLoading(false);
+      setLoadingTodosIds(prevTodos => prevTodos.filter(todoId => todoId !== 0));
+    }
+  };
+
+  const removeTodo = async (id: number) => {
+    setIsLoading(true);
+    setLoadingTodosIds(prev => [...prev, id]);
+
+    try {
+      await deleteTodo(id);
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== id));
+    } catch {
+      setError(Error.UNABLE_TO_DELETE);
+    } finally {
+      setIsLoading(false);
+      setLoadingTodosIds(prevTodos =>
+        prevTodos.filter(todoId => todoId !== id),
+      );
+    }
+  };
+
+  const changeTodoStatus = async (id: number, status: boolean) => {
+    setIsLoading(true);
+    setLoadingTodosIds(prev => [...prev, id]);
+
+    const preparedForUpdate = todos.find(todo => todo.id === id);
+
+    if (preparedForUpdate) {
+      try {
+        await toggleTodoStatus(id, status);
+        setTodos(prevTodos =>
+          prevTodos.map(prev =>
+            prev.id === id ? { ...prev, completed: status } : prev,
+          ),
+        );
+      } catch {
+        setError(Error.UNABLE_TO_UPDATE);
+      } finally {
+        setIsLoading(false);
+        setLoadingTodosIds(prev => prev.filter(todoId => todoId !== id));
+      }
+    }
+  };
+
+  const toggleOne = (id: number) => {
+    const preparedForUpdate = todos.find(todo => todo.id === id);
+
+    if (preparedForUpdate) {
+      changeTodoStatus(id, !preparedForUpdate.completed);
+    }
+  };
+
+  const toggleAll = (status: boolean) => {
+    todos.forEach(({ id, completed }) => {
+      if (completed !== status) {
+        changeTodoStatus(id, status);
+      }
+    });
+  };
+
+  const updateTitle = async (
+    id: number,
+    newTitle: string,
+    setEditTodoId: (id: number | null) => {},
+  ) => {
```

Comment:
I do not really like this `setEditingTodoId` arg, because it is not related to `updateTitle` logic it's more about specific UI.

Can we return Promise from this function and then just handle it in our component? 

If it's not the solution, rename `setEditTodoId` to something more generic like `callback`.